h2o-core/src/main/java/water/Key.java

```diff
@@ -255,14 +252,29 @@ private Key(byte[] kb) {
     hash += (hash << 3);
     hash ^= (hash >> 11);
     hash += (hash << 15);
+    return hash;
+  }
+
+  // Construct a new Key.
+  private Key(byte[] kb) {
+    this(kb, hash(kb));
+  }
+
+  private Key(byte[] kb, int hash) {
+    if( kb.length > KEY_LENGTH ) throw new IllegalArgumentException("Key length would be "+kb.length);
+    _kb = kb;
     _hash = hash;
   }
 
+  static <T extends Keyed> Key<T> buildKeyForTestingPurposes(byte[] kb, int hash) {
+    return new Key<>(kb, hash);
+  }
+
   // Make new Keys.  Optimistically attempt interning, but no guarantee.
   static <P extends Keyed> Key<P> make(byte[] kb, byte rf) {
     if( rf == -1 ) throw new IllegalArgumentException();
-    Key key = new Key(kb);
-    Key key2 = H2O.getk(key); // Get the interned version, if any
+    Key<P> key = new Key<>(kb.clone());
```

Comment:
For an arbitrary `byte[]` array making a clone is the right thing to do. However often we make a Key from a String (which is then cast to byte[]) -- there should be a version of the constructor that optimizes that use case.